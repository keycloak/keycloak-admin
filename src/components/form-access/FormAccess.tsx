import React, {
  Children,
  cloneElement,
  isValidElement,
  ReactElement,
} from "react";
import { Controller } from "react-hook-form";
import { useTranslation } from "react-i18next";
import { Form, FormProps, Switch } from "@patternfly/react-core";

import { useAccess } from "../../context/access/Access";
import { AccessType } from "../../context/whoami/who-am-i-model";

export type FormAccessProps = FormProps & {
  role: AccessType;
  fineGrainedAccess?: boolean;
  unWrap?: boolean;
  children: ReactElement[];
};

export const FormAccess = ({
  children,
  role,
  fineGrainedAccess = false,
  unWrap = false,
  ...rest
}: FormAccessProps) => {
  const { hasAccess } = useAccess();

  const ReadOnly = (props: any) => {
    const { t } = useTranslation();
    if (typeof props.value === "boolean") {
      return (
        <Switch
          id={props.name}
          isDisabled
          label={t("common:on")}
          labelOff={t("common:off")}
          isChecked={props.value}
          onChange={props.onChange}
        />
      );
    }
    return (
      <input
        type="text"
        value={props.value}
        onChange={props.onChange}
        disabled
      />
    );
  };

  const recursiveCloneChildren = (children: ReactElement[], newProps: any) => {
    return Children.map(children, (child: ReactElement) => {
      if (!isValidElement(child)) {
        return child;
      }

      if (child.props) {
        if (child.type === Controller && newProps.isDisabled) {
          return cloneElement(child, {
            ...(child as ReactElement).props,
            // eslint-disable-next-line react/display-name
            render: (props: any) => (
              <ReadOnly {...props} name={(child as ReactElement).props.name} />
            ),
          });
        }
        newProps.children = recursiveCloneChildren(
          (child as ReactElement).props.children,
          newProps
        );
        return cloneElement(child, newProps);
      }
      return child;
    });
  };
  return (
    <>
      {!unWrap && (
        <Form {...rest}>
          {recursiveCloneChildren(children, {
            isDisabled: !hasAccess(role) && !fineGrainedAccess,
            readOnly: !hasAccess(role) && !fineGrainedAccess,
          })}
        </Form>
      )}
      {unWrap &&
        recursiveCloneChildren(children, {
          isDisabled: !hasAccess(role) && !fineGrainedAccess,
          readOnly: !hasAccess(role) && !fineGrainedAccess,
        })}
    </>
  );
};
