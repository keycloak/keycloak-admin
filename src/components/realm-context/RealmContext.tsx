import React, { useState, useContext } from "react";
import { WhoAmIContext } from "../../whoami/WhoAmI";

export const RealmContext = React.createContext({
  realm: "",
  setRealm: (realm: string) => {},
});

type RealmContextProviderProps = { children: React.ReactNode };

export const RealmContextProvider = ({
  children,
}: RealmContextProviderProps) => {
  const homeRealm = useContext(WhoAmIContext).getHomeRealm();
  const [realm, setRealm] = useState(homeRealm);

  return (
    <RealmContext.Provider value={{ realm, setRealm }}>
      {children}
    </RealmContext.Provider>
  );
};
