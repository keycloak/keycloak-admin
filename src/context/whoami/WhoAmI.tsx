import React, { useContext, useEffect, useState } from "react";
import { useErrorHandler } from "react-error-boundary";
import i18n from "../../i18n";

import { asyncStateFetch, useAdminClient } from "../auth/AdminClient";
import { RealmContext } from "../realm-context/RealmContext";
import WhoAmIRepresentation, {
  AccessType,
} from "keycloak-admin/lib/defs/whoAmIRepresentation";

export class WhoAmI {
  constructor(
    private homeRealm?: string | undefined,
    private me?: WhoAmIRepresentation | undefined
  ) {
    if (this.me !== undefined && this.me.locale) {
      i18n.changeLanguage(this.me.locale, (error) => {
        if (error) console.error("Unable to set locale to", this.me?.locale);
      });
    }
  }

  public getDisplayName(): string {
    if (this.me === undefined) return "";

    return this.me.displayName;
  }

  /**
   * Return the realm I am signed in to.
   */
  public getHomeRealm(): string {
    let realm: string | undefined = this.homeRealm;
    if (realm === undefined) realm = this.me?.realm;
    if (realm === undefined) realm = "master"; // this really can't happen in the real world

    return realm;
  }

  public canCreateRealm(): boolean {
    return this.me !== undefined && this.me.createRealm;
  }

  public getRealmAccess(): Readonly<{
    [key: string]: ReadonlyArray<AccessType>;
  }> {
    if (this.me === undefined) return {};

    return this.me.realm_access;
  }
}

type WhoAmIProps = {
  refresh: () => void;
  whoAmI: WhoAmI;
};

export const WhoAmIContext = React.createContext<WhoAmIProps>({
  refresh: () => {},
  whoAmI: new WhoAmI(),
});

type WhoAmIProviderProps = { children: React.ReactNode };
export const WhoAmIContextProvider = ({ children }: WhoAmIProviderProps) => {
  const adminClient = useAdminClient();
  const handleError = useErrorHandler();
  const { realm, setRealm } = useContext(RealmContext);
  const [whoAmI, setWhoAmI] = useState<WhoAmI>(new WhoAmI());
  const [key, setKey] = useState(0);

  useEffect(() => {
    return asyncStateFetch(
      () => adminClient.whoAmI.find(),
      (me) => {
        const whoAmI = new WhoAmI(adminClient.keycloak?.realm, me);
        if (!realm) {
          setRealm(whoAmI.getHomeRealm());
        }
        setWhoAmI(whoAmI);
      },
      handleError
    );
  }, [key]);

  return (
    <WhoAmIContext.Provider value={{ refresh: () => setKey(key + 1), whoAmI }}>
      {children}
    </WhoAmIContext.Provider>
  );
};
