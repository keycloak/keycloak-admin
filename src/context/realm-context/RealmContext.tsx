import React, { useContext, useEffect, useState } from "react";
import _ from "lodash";

import RealmRepresentation from "keycloak-admin/lib/defs/realmRepresentation";
import { RecentUsed } from "../../components/realm-selector/recent-used";
import { useErrorHandler } from "react-error-boundary";
import { asyncStateFetch, useAdminClient } from "../auth/AdminClient";
import { WhoAmIContext } from "../whoami/WhoAmI";

type RealmContextType = {
  realm: string;
  setRealm: (realm: string) => void;
  realms: RealmRepresentation[];
  refresh: () => Promise<void>;
};

export const RealmContext = React.createContext<RealmContextType>({
  realm: "",
  setRealm: () => {},
  realms: [],
  refresh: () => Promise.resolve(),
});

type RealmContextProviderProps = { children: React.ReactNode };

export const RealmContextProvider = ({
  children,
}: RealmContextProviderProps) => {
  const { whoAmI } = useContext(WhoAmIContext);
  const [realm, setRealm] = useState(whoAmI.getHomeRealm());
  const [realms, setRealms] = useState<RealmRepresentation[]>([]);
  const adminClient = useAdminClient();
  const errorHandler = useErrorHandler();
  const recentUsed = new RecentUsed();

  const updateRealmsList = (realms: RealmRepresentation[]) => {
    setRealms(_.sortBy(realms, "realm"));
    recentUsed.clean(realms.map((r) => r.realm!));
  };

  useEffect(
    () =>
      asyncStateFetch(
        () => adminClient.realms.find(),
        (realms) => updateRealmsList(realms),
        errorHandler
      ),
    []
  );

  const set = (realm: string) => {
    if (
      realms.length === 0 ||
      realms.findIndex((r) => r.realm == realm) !== -1
    ) {
      recentUsed.setRecentUsed(realm);
      setRealm(realm);
      adminClient.setConfig({
        realmName: realm,
      });
    }
  };
  return (
    <RealmContext.Provider
      value={{
        realm,
        setRealm: set,
        realms,
        refresh: async () => {
          const list = await adminClient.realms.find();
          updateRealmsList(list);
        },
      }}
    >
      {children}
    </RealmContext.Provider>
  );
};

export const useRealm = () => useContext(RealmContext);
