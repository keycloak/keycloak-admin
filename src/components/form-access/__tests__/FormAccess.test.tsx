import React from "react";
import { mount } from "enzyme";
import { Controller, useForm } from "react-hook-form";
import { FormGroup, Switch, TextInput } from "@patternfly/react-core";
import type WhoAmIRepresentation from "keycloak-admin/lib/defs/whoAmIRepresentation";

import { WhoAmI, WhoAmIContext } from "../../../context/whoami/WhoAmI";
import whoami from "../../../context/whoami/__tests__/mock-whoami.json";
import { RealmContext } from "../../../context/realm-context/RealmContext";
import { AccessContextProvider } from "../../../context/access/Access";

import { FormAccess } from "../FormAccess";

describe("<FormAccess />", () => {
  const Form = ({ realm }: { realm: string }) => {
    const { register, control } = useForm();
    return (
      <WhoAmIContext.Provider
        value={{
          refresh: () => {},
          whoAmI: new WhoAmI("master", whoami as WhoAmIRepresentation),
        }}
      >
        <RealmContext.Provider
          value={{
            realm,
            setRealm: () => {},
            realms: [],
            refresh: () => Promise.resolve(),
          }}
        >
          <AccessContextProvider>
            <FormAccess role="manage-clients">
              <FormGroup label="test" fieldId="field">
                <TextInput type="text" id="field" {...register("fieldName")} />
              </FormGroup>
              <Controller
                name="consentRequired"
                defaultValue={false}
                control={control}
                render={({ onChange, value }) => (
                  <Switch
                    id="kc-consent"
                    label={"on"}
                    labelOff="off"
                    isChecked={value}
                    onChange={onChange}
                  />
                )}
              />
            </FormAccess>
          </AccessContextProvider>
        </RealmContext.Provider>
      </WhoAmIContext.Provider>
    );
  };
  it("render normal form", () => {
    const comp = mount(<Form realm="master" />);
    expect(comp).toMatchSnapshot();
  });

  it("render form disabled for test realm", () => {
    const container = mount(<Form realm="test" />);

    const disabled = container.find("input#field").props().disabled;
    expect(disabled).toBe(true);

    expect(container.find("input#kc-consent").props().disabled).toBe(true);
  });
});
