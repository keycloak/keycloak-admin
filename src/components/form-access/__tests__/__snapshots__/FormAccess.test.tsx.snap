// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormAccess /> render normal form 1`] = `
<Form
  realm="master"
>
  <AccessContextProvider>
    <FormAccess
      role="manage-clients"
    >
      <Form
        className="keycloak__form "
      >
        <form
          className="pf-c-form keycloak__form "
          noValidate={true}
        >
          <FormGroup
            fieldId="field"
            key=".0"
            label="test"
          >
            <div
              className="pf-c-form__group"
            >
              <div
                className="pf-c-form__group-label"
              >
                <label
                  className="pf-c-form__label"
                  htmlFor="field"
                >
                  <span
                    className="pf-c-form__label-text"
                  >
                    test
                  </span>
                </label>
                 
              </div>
              <div
                className="pf-c-form__group-control"
              >
                <TextInput
                  id="field"
                  key=".0"
                  name="fieldName"
                  type="text"
                >
                  <TextInputBase
                    aria-label={null}
                    className=""
                    id="field"
                    innerRef={[Function]}
                    isDisabled={false}
                    isLeftTruncated={false}
                    isReadOnly={false}
                    isRequired={false}
                    name="fieldName"
                    onChange={[Function]}
                    type="text"
                    validated="default"
                  >
                    <input
                      aria-invalid={false}
                      aria-label={null}
                      className="pf-c-form-control"
                      disabled={false}
                      id="field"
                      name="fieldName"
                      onBlur={[Function]}
                      onChange={[Function]}
                      onFocus={[Function]}
                      readOnly={false}
                      required={false}
                      type="text"
                    />
                  </TextInputBase>
                </TextInput>
              </div>
            </div>
          </FormGroup>
          <Controller
            control={
              Object {
                "defaultValuesRef": Object {
                  "current": Object {},
                },
                "fieldArrayDefaultValuesRef": Object {
                  "current": Object {},
                },
                "fieldArrayNamesRef": Object {
                  "current": Set {},
                },
                "fieldArrayValuesRef": Object {
                  "current": Object {},
                },
                "fieldsRef": Object {
                  "current": Object {
                    "consentRequired": Object {
                      "ref": Object {
                        "focus": [Function],
                        "name": "consentRequired",
                      },
                    },
                    "fieldName": Object {
                      "ref": <input
                        aria-invalid="false"
                        class="pf-c-form-control"
                        id="field"
                        name="fieldName"
                        type="text"
                        value=""
                      />,
                    },
                  },
                },
                "fieldsWithValidationRef": Object {
                  "current": Object {},
                },
                "formState": Object {
                  "dirtyFields": Object {},
                  "errors": Object {},
                  "isDirty": false,
                  "isSubmitSuccessful": false,
                  "isSubmitted": false,
                  "isSubmitting": false,
                  "isValid": false,
                  "isValidating": false,
                  "submitCount": 0,
                  "touched": Object {},
                },
                "formStateRef": Object {
                  "current": Object {
                    "dirtyFields": Object {},
                    "errors": Object {},
                    "isDirty": false,
                    "isSubmitSuccessful": false,
                    "isSubmitted": false,
                    "isSubmitting": false,
                    "isValid": false,
                    "isValidating": false,
                    "submitCount": 0,
                    "touched": Object {},
                  },
                },
                "getValues": [Function],
                "isFormDirty": [Function],
                "mode": Object {
                  "isOnAll": false,
                  "isOnBlur": false,
                  "isOnChange": false,
                  "isOnSubmit": true,
                  "isOnTouch": false,
                },
                "reValidateMode": Object {
                  "isReValidateOnBlur": false,
                  "isReValidateOnChange": true,
                },
                "readFormStateRef": Object {
                  "current": Object {
                    "constructor": true,
                    "dirtyFields": true,
                    "errors": true,
                    "isDirty": true,
                    "isSubmitSuccessful": true,
                    "isSubmitted": true,
                    "isSubmitting": true,
                    "isValid": true,
                    "isValidating": true,
                    "submitCount": true,
                    "touched": true,
                  },
                },
                "register": [Function],
                "removeFieldEventListener": [Function],
                "resetFieldArrayFunctionRef": Object {
                  "current": Object {},
                },
                "setValue": [Function],
                "shallowFieldsStateRef": Object {
                  "current": Object {},
                },
                "shouldUnregister": true,
                "trigger": [Function],
                "unregister": [Function],
                "updateFormState": [Function],
                "updateWatchedValue": [Function],
                "useWatchFieldsRef": Object {
                  "current": Object {},
                },
                "useWatchRenderFunctionsRef": Object {
                  "current": Object {},
                },
                "validFieldsRef": Object {
                  "current": Object {},
                },
                "validateResolver": undefined,
                "watchInternal": [Function],
              }
            }
            defaultValue={false}
            key=".1"
            name="consentRequired"
            render={[Function]}
          >
            <Switch
              aria-label=""
              id="kc-consent"
              isChecked={false}
              isDisabled={false}
              label="on"
              labelOff="off"
              onChange={[Function]}
            >
              <label
                className="pf-c-switch"
                data-ouia-component-id="OUIA-Generated-Switch-1"
                data-ouia-component-type="PF4/Switch"
                data-ouia-safe={true}
                htmlFor="kc-consent"
              >
                <input
                  aria-label=""
                  aria-labelledby="kc-consent-on"
                  checked={false}
                  className="pf-c-switch__input"
                  disabled={false}
                  id="kc-consent"
                  onChange={[Function]}
                  type="checkbox"
                />
                <span
                  className="pf-c-switch__toggle"
                />
                <span
                  aria-hidden="true"
                  className="pf-c-switch__label pf-m-on"
                  id="kc-consent-on"
                >
                  on
                </span>
                <span
                  aria-hidden="true"
                  className="pf-c-switch__label pf-m-off"
                  id="kc-consent-off"
                >
                  off
                </span>
              </label>
            </Switch>
          </Controller>
        </form>
      </Form>
    </FormAccess>
  </AccessContextProvider>
</Form>
`;
